javascript/ql/src/semmle/javascript/DOM.qll

```diff
@@ -298,6 +298,23 @@ module DOM {
       )
     }
 
+    /**
+     * A data flow node that might refer to some form.
+     * Either by a read like `document.forms[0]`, or a property read from `document` with some constant property-name.
+     * E.g. if `<form name="foobar">..</form>` exists, then `document.foobar` refers to that form.
+     */
+    private DataFlow::SourceNode forms() {
+      result = documentRef().getAPropertyRead("forms").getAPropertyRead()
+      or
+      exists(DataFlow::PropRead read |
+        read = documentRef().getAPropertyRead() and
+        result = read
+      |
+        read.mayHavePropertyName(_) and
+        not read.mayHavePropertyName(getADomPropertyName())
```

Comment:
With the addition of this rule it seems we now consider all properties of `document` to be DOM value sources.

Maybe we should just express that in a more direct way.

Reply:
Are you suggesting that I put `documentRef().getAPropertyRead(any(string s | not s = getADomPropertyName()))` into `DefaultRange`? 

I'm also using the `forms` predicate further down, so I would also like to keep the prop-read inside the `forms` predicate.

Reply:
I'm just saying that all of `documentRef().getAPropertyRead()` is now a DOM value source, but it's done in a very non-obvious, roundabout way, by unioning three sets:
- all prop reads with a DOM property name
- all prop reads with a non-DOM property name, and
- all property reads with an unknown property name.

If we end up with all of them anyway, I'd just prefer to have `documentRef().getAPropertyRead()` be listed as an explicit case in `DefaultRange`. It's fine to keep the `forms()` predicate.

Reply:
> documentRef().getAPropertyRead() is now a DOM value source 

Hmmm. That turns out not to be the case.  
I tried to add `documentRef().getAPropertyRead()`, but then the tests failed because `document.location` became a `domValueRef`. 

`domValueRef()` does not contain `documentRef()`, so the property-read near the top of `DefaultRange' does not refer to `document`.